### PR TITLE
Add simple event recorder

### PR DESCRIPTION
**Pull request by mfojtik**
Adding event recorder interface for recording events from operators. Recorder should be wired to resourceapply package and observer package to. Events published by operator are for historical action debugging.

This implementation require the ObjectReference to be passed to NewRecorder in the factory when the operator is started. It means that the `EVENT_SOURCE_REPLICASET_NAME` will have to be passed via downward API into the container running the operator and the replica set is then used as the object we recording the events against. These events should be also propagated to the owning deployment.

/cc @deads2k 
/cc @sttts

### CODE REVIEW

File: pkg/operator/events/recorder.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+func (r recorder) makeEvent(source RecorderSource, eventType, reason, message string) *corev1.Event {
+	currentTime := metav1.Time{Time: time.Now()}
+	gvk := source.GroupVersionKind()
+	apiVersion, kind := gvk.ToAPIVersionAndKind()
+
+	event := &corev1.Event{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%v.%x", source.GetObjectMeta().GetName(), currentTime.UnixNano()),
```
Comment (sttts) [inline pkg/operator/events/recorder.go]: never looked into how events are named. this is how? 😱

Comment (mfojtik) [reply pkg/operator/events/recorder.go]: yes, copied from deployment controller :)

File: pkg/operator/events/recorder_deployment.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+var eventSourceReplicaSetNameEnvFunc = func() string {
+	return os.Getenv(eventSourceReplicaSetNameEnv)
+}
+
+// DeploymentFromEnvRecorderSource return event recorder source using the EVENT_SOURCE_REPLICASET_NAME environment variable
+// set inside the container created by deployment.
+func DeploymentFromEnvRecorderSource(client appsv1client.ReplicaSetInterface) (RecorderSourceFunc, error) {
+	replicaSetName := eventSourceReplicaSetNameEnvFunc()
+	if len(replicaSetName) == 0 {
+		return nil, fmt.Errorf("Unable to setup event recorder as '%s' env variable is not set.", eventSourceReplicaSetNameEnv)
```
Comment (sttts) [inline pkg/operator/events/recorder_deployment.go]: lower case error

Comment (sttts) [reply pkg/operator/events/recorder_deployment.go]: %q instead of quoted %s

File: pkg/operator/events/recorder_deployment.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+import (
+	"fmt"
+	"os"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	appsv1client "k8s.io/client-go/kubernetes/typed/apps/v1"
+)
+
+const eventSourceReplicaSetNameEnv = "EVENT_SOURCE_REPLICASET_NAME"
```
Comment (sttts) [inline pkg/operator/events/recorder_deployment.go]: document what this does

File: pkg/operator/events/recorder.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+	CreateEvent(reason, message string) error
+	CreateEventWarning(reason, message string) error
+}
+
+type RecorderSource interface {
+	GroupVersionKind() schema.GroupVersionKind
+	metav1.ObjectMetaAccessor
+}
+
+func NewRecorder(client corev1client.EventInterface, sourceComponent string, sourceFunc RecorderSourceFunc) Recorder {
```
Comment (sttts) [inline pkg/operator/events/recorder.go]: why the RecorderSourceFunc and not RecorderSource directly? The func without arguments is like a constant for the record, isn't it?

File: pkg/operator/events/recorder.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+type RecorderSourceFunc func() (RecorderSource, error)
```
Comment (soltysh) [inline pkg/operator/events/recorder.go]: Why not use `k8s.io/client-go/tools/record` package instead of writing your own, we already vendor it

Comment (mfojtik) [reply pkg/operator/events/recorder.go]: Because there is no way to only use the EventRecorder without adding EventBroadcaster. Because the EventRecorder does not set all the events fields we want to set and I really want to be in control what fields are set and how. Because we will have to create an instance of this recorder everything we creating an event to fill the EventSource (as you can't dynamically change the involved object, etc..)

In summary, I'm not a fan of re-using things if they don't fit this use case. Especially if this recorder does not need to be super sophisticated and we don't really need/want the sophistication of EventSink.

File: pkg/operator/events/recorder.go (hunk 4 above)
Comment (mfojtik) [reply pkg/operator/events/recorder.go]: yeah, good point

File: pkg/operator/events/recorder.go (hunk 5 above)
Comment (soltysh) [reply pkg/operator/events/recorder.go]: Discussed this in length on Slack, I'm ok with it for now :wink:

File: pkg/operator/events/recorder.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+func NewRecorder(client corev1client.EventInterface, sourceComponent string, sourceFunc RecorderSourceFunc) Recorder {
+	return &recorder{
+		eventClient:     client,
+		sourceFunc:      sourceFunc,
+		sourceComponent: sourceComponent,
+	}
+}
+
+type recorderSourceObject struct {
```
Comment (soltysh) [inline pkg/operator/events/recorder.go]: Why this? Why not just using `ObjectReference`, all you care is there already? This also applies to the interface `RecordSource`, why not just use `runtime.Object`?

File: pkg/operator/events/recorder.go (hunk 4 above)
Comment (mfojtik) [reply pkg/operator/events/recorder.go]: @sttts actually, `RecorderSourceFunc` returns the RecorderSource and error and it allows things like `DeploymentFromEnvRecorderSource` to be used as argument. I don't see how to refactor this to loose the RecorderSourceFunc ?

File: pkg/operator/events/recorder.go (hunk 5 above)
Comment (soltysh) [reply pkg/operator/events/recorder.go]: How about instead having your own implementation of `EventRecorder` interface, this will save us some work in the future, if we need to switch?

File: pkg/operator/events/recorder.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+	eventClient     corev1client.EventInterface
+	sourceFunc      RecorderSourceFunc
+	sourceComponent string
+}
+
+// Event emits the normal type event.
+func (r *recorder) Event(reason, message string) error {
+	source, err := r.sourceFunc()
+	if err != nil {
+		return fmt.Errorf("enable to get event source: %v", err)
```
Comment (soltysh) [inline pkg/operator/events/recorder.go]: s/enable/unable

Comment (soltysh) [reply pkg/operator/events/recorder.go]: or failed

File: pkg/operator/events/recorder.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+	// This will likely never happen.
+	return fmt.Errorf("unable to create event as the source object is not specified")
+}
+
+// EventWarning emits the warning type event.
+func (r *recorder) EventWarning(reason, message string) error {
+	source, err := r.sourceFunc()
+	if err != nil {
+		return fmt.Errorf("enable to get event source: %v", err)
```
Comment (soltysh) [inline pkg/operator/events/recorder.go]: ditto

Comment (soltysh) [reply pkg/operator/events/recorder.go]: Why not move this block into `makeEvent`?

File: pkg/operator/events/recorder_deployment.go [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+	if len(replicaSetName) == 0 {
+		return nil, fmt.Errorf("unable to setup event recorder as %q env variable is not set", eventSourceReplicaSetNameEnv)
+	}
+
+	return func() (*corev1.ObjectReference, error) {
+		rs, err := client.Get(replicaSetName, metav1.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+		return &corev1.ObjectReference{
```
Comment (soltysh) [inline pkg/operator/events/recorder_deployment.go]: Use `k8s.io/client-go/tools/reference#GetReference` instead

File: pkg/operator/events/recorder_deployment.go [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+const eventSourceReplicaSetNameEnv = "EVENT_SOURCE_REPLICASET_NAME"
+
+// eventSourceReplicaSetNameEnvFunc allows to override the way we get the environment variable value (for unit tests).
+var eventSourceReplicaSetNameEnvFunc = func() string {
+	return os.Getenv(eventSourceReplicaSetNameEnv)
+}
+
+// DeploymentFromEnvRecorderSource return event recorder source using the EVENT_SOURCE_REPLICASET_NAME environment variable
+// set inside the container created by deployment.
+// TODO: Switching to a lister will make this faster.
```
Comment (sttts) [inline pkg/operator/events/recorder_deployment.go]: it's called once per process, isn't  it?

Comment (mfojtik) [reply pkg/operator/events/recorder_deployment.go]: yeah, this todo is not relevant
